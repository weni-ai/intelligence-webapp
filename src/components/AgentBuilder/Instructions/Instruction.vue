<template>
  <section class="instruction">
    <p
      class="instruction__text"
      data-testid="instruction-text"
    >
      {{ text }}
    </p>

    <p
      v-if="tag"
      class="instruction__tag"
      data-testid="instruction-tag"
    >
      {{ tag }}
    </p>
  </section>
</template>

<script setup>
const props = defineProps({
  text: {
    type: String,
    required: true,
  },
  tag: {
    type: String,
    default: '',
    required: false,
  },
});
</script>

<style lang="scss" scoped>
.instruction {
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: $unnnic-spacing-sm;

  padding: $unnnic-spacing-sm;

  &:not(:last-child) {
    border-bottom: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
  }

  &__text {
    margin: 0;
    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__tag {
    border-radius: $unnnic-border-radius-pill;

    padding: $unnnic-spacing-nano $unnnic-spacing-ant;

    background-color: $unnnic-color-neutral-light;
    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-md;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
  }
}
</style>
