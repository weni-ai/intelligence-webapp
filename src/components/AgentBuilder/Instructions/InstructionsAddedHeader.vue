<template>
  <header class="instructions-added-header">
    <hgroup
      class="instructions-added-header__infos"
      data-testid="instructions-header-infos"
    >
      <h2
        class="infos__title"
        data-testid="instructions-title"
      >
        {{ $t('agent_builder.instructions.instructions_added.title') }}
      </h2>
      <h3
        class="infos__subtitle"
        data-testid="instructions-subtitle"
      >
        {{
          $t('agent_builder.instructions.instructions_added.subtitle', {
            count: instructionsCount,
          })
        }}
      </h3>
    </hgroup>

    <UnnnicSelectSmart
      v-model="filters"
      :options="filtersOptions"
      orderedByIndex
      multiple
      multipleWithoutSelectsMessage
      autocomplete
      data-testid="instructions-filters"
    />
  </header>
</template>

<script setup>
import { computed } from 'vue';
import i18n from '@/utils/plugins/i18n';

defineProps({
  instructionsCount: {
    type: Number,
    required: true,
  },
});

const filters = defineModel('filters', {
  type: Array,
  required: true,
});

const getFilterTranslation = (filter) =>
  i18n.global.t(
    `agent_builder.instructions.instructions_added.filters.${filter}`,
  );
const filtersOptions = computed(() => [
  {
    label: getFilterTranslation('placeholder'),
    value: '',
  },
  {
    label: getFilterTranslation('default_instructions'),
    value: 'default_instructions',
  },
  {
    label: getFilterTranslation('personalized_instructions'),
    value: 'personalized_instructions',
  },
]);
</script>

<style lang="scss" scoped>
.instructions-added-header {
  display: grid;
  grid-template-columns: 9fr 3fr;
  align-items: center;
  gap: $unnnic-spacing-sm;

  &__infos {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-nano;

    .infos__title {
      margin: 0;

      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-bold;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }

    .infos__subtitle {
      margin: 0;

      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }
  }
}
</style>
