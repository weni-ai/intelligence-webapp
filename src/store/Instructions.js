import { reactive, ref, computed } from 'vue';
import { defineStore } from 'pinia';

import { useProjectStore } from './Project';
import { useAlertStore } from './Alert';

import nexusaiAPI from '@/api/nexusaiAPI';

import i18n from '@/utils/plugins/i18n';

function callAlert(type, alertText) {
  const alertStore = useAlertStore();
  alertStore.add({
    text: i18n.global.t(`agent_builder.instructions.${alertText}`),
    type,
  });
}

export const useInstructionsStore = defineStore('Instructions', () => {
  const projectUuid = computed(() => useProjectStore().uuid);

  const instructions = reactive({
    data: [],
    status: null,
  });

  const newInstruction = reactive({
    text: '',
    status: null,
  });

  const activeInstructionsListTab = ref('custom');

  async function addInstruction() {
    newInstruction.status = 'loading';
    activeInstructionsListTab.value = 'custom';

    try {
      const instructionResponse =
        await nexusaiAPI.agent_builder.instructions.addInstruction({
          projectUuid: projectUuid.value,
          instruction: newInstruction,
        });

      instructions.data.unshift({
        ...newInstruction,
        id: instructionResponse.id,
      });
      newInstruction.status = null;
      newInstruction.text = '';

      callAlert('success', 'new_instruction.success_alert');
    } catch (error) {
      newInstruction.status = 'error';
      callAlert('error', 'new_instruction.error_alert');
    }
  }

  async function loadInstructions() {
    instructions.status = 'loading';
    try {
      const response = await nexusaiAPI.agent_builder.instructions.list({
        projectUuid: projectUuid.value,
      });

      instructions.data = [...instructions.data, ...response];
      instructions.status = 'complete';
    } catch (error) {
      instructions.status = 'error';
    }
  }
  async function editInstruction(id, text) {
    const instruction = instructions.data.find(
      (instruction) => instruction.id === id,
    );

    if (!instruction) return;

    try {
      instruction.status = 'loading';
      await nexusaiAPI.agent_builder.instructions.edit({
        projectUuid: projectUuid.value,
        id,
        text,
      });
      instruction.text = text;
      instruction.status = 'complete';

      callAlert('success', 'edit_instruction.success_alert');
    } catch (error) {
      instruction.status = 'error';
      callAlert('error', 'edit_instruction.error_alert');
    }

    return { status: instruction.status };
  }

  async function removeInstruction(id) {
    const instruction = instructions.data.find(
      (instruction) => instruction.id === id,
    );
    if (!instruction) return;
    instruction.status = 'loading';

    try {
      await nexusaiAPI.agent_builder.instructions.delete({
        projectUuid: projectUuid.value,
        id,
      });
      instructions.data = instructions.data.filter(
        (instruction) => instruction.id !== id,
      );
      callAlert('default', 'remove_instruction.success_alert');
    } catch (error) {
      instruction.status = 'error';
      callAlert('error', 'remove_instruction.error_alert');
    }

    return { status: instruction.status };
  }

  return {
    instructions,
    newInstruction,
    activeInstructionsListTab,
    addInstruction,
    loadInstructions,
    editInstruction,
    removeInstruction,
  };
});
