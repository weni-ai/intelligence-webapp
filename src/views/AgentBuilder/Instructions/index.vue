<template>
  <section class="instructions">
    <AgentBuilderHeader
      data-testid="instructions-header"
      :withDivider="false"
      actionsSize="none"
    />
    <NewInstruction data-testid="new-instruction" />
  </section>
</template>

<script setup>
import AgentBuilderHeader from '@/components/AgentBuilder/Header.vue';
import NewInstruction from '@/components/AgentBuilder/Instructions/NewInstruction.vue';
</script>

<style lang="scss" scoped>
.instructions {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-md;
}
</style>
