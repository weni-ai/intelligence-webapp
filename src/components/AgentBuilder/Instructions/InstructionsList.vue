<template>
  <section
    :class="[
      'instructions-list',
      {
        'instructions-list--loading': isLoading,
        'instructions-list--no-instructions': instructions.length === 0,
      },
    ]"
    data-testid="instructions"
  >
    <template v-if="isLoading">
      <UnnnicSkeletonLoading
        v-for="index in 10"
        :key="index"
        tag="div"
        width="100%"
        height="60px"
        data-testid="instruction-loading"
      />
    </template>
    <section
      v-else-if="instructions.length === 0"
      class="instructions-list__no-instructions"
    >
      <UnnnicIcon
        icon="format_list_bulleted"
        size="xl"
        scheme="neutral-cleanest"
      />
      <p class="no-instructions__text">
        {{
          $t(
            'agent_builder.instructions.instructions_list.no_custom_instructions',
          )
        }}
      </p>
    </section>
    <template v-else>
      <Instruction
        v-for="instruction in instructions"
        :key="instruction.id"
        :instruction="instruction"
        :showActions="showActions"
        data-testid="instruction-added"
      />
    </template>
  </section>
</template>

<script setup>
import Instruction from './Instruction.vue';

defineProps({
  instructions: {
    type: Array,
    required: true,
  },
  isLoading: {
    type: Boolean,
    default: false,
  },
  showActions: {
    type: Boolean,
    default: false,
  },
});
</script>

<style lang="scss" scoped>
.instructions-list {
  margin-bottom: $unnnic-spacing-sm;

  display: flex;
  flex-direction: column;

  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

  &--loading,
  &--no-instructions {
    height: 100%;

    gap: $unnnic-spacing-nano;

    border: none;
    border-radius: 0;
  }

  &__no-instructions {
    height: 100%;

    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: $unnnic-spacing-nano;

    .no-instructions__text {
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-gt;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }
  }
}
</style>
