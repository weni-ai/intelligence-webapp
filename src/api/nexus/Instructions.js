import nexusRequest from '../nexusaiRequest';
import { InstructionAdapter } from '../adapters/instructions/instruction';
import { useInstructionsStore } from '@/store/Instructions';

const request = nexusRequest;

export const Instructions = {
  async addInstruction({ projectUuid, instruction }) {
    const instructionsStore = useInstructionsStore();
    const body = {
      instructions: [...instructionsStore.instructions.data, instruction].map(
        InstructionAdapter.toApi,
      ),
    };
    const response = await request.$http.put(
      `api/${projectUuid}/customization/`,
      body,
      {
        hideGenericErrorAlert: true,
      },
    );

    const biggestId = response.data.instructions.reduce((acc, instruction) => {
      if (instruction.id > acc) acc = instruction.id;
      return acc;
    }, 0);

    return InstructionAdapter.fromApi(
      response.data.instructions.find(
        (instruction) => instruction.id === biggestId,
      ),
    );
  },

  async list({ projectUuid }) {
    const response = await request.$http.get(
      `api/${projectUuid}/customization/`,
    );

    return response.data.instructions.map(InstructionAdapter.fromApi);
  },

  async edit({ projectUuid, id, text }) {
    const instructionsStore = useInstructionsStore();

    const body = {
      instructions: [
        ...instructionsStore.instructions.data.filter(
          (instruction) => instruction.id !== id,
        ),
        { id, text },
      ].map(InstructionAdapter.toApi),
    };
    await request.$http.put(`api/${projectUuid}/customization/`, body);
  },

  async delete({ projectUuid, id }) {
    await request.$http.delete(`api/${projectUuid}/customization/?id=${id}`);
  },
};
