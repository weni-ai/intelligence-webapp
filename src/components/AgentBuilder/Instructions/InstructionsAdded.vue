<template>
  <section class="instructions-added">
    <h2
      class="instructions-added__title"
      data-testid="instructions-title"
    >
      {{ $t('agent_builder.instructions.instructions_added.title') }}
    </h2>

    <section
      :class="[
        'instructions-added__instructions',
        {
          'instructions-added__instructions--loading':
            instructionsStore.instructions.status === 'loading',
        },
      ]"
      data-testid="instructions"
    >
      <template v-if="instructionsStore.instructions.status === 'loading'">
        <UnnnicSkeletonLoading
          v-for="index in 10"
          :key="index"
          tag="div"
          width="100%"
          height="60px"
          data-testid="instruction-loading"
        />
      </template>
      <template v-else>
        <Instruction
          v-for="(instruction, index) in instructionsDefault"
          :key="index"
          data-testid="instruction-default"
          :text="instruction"
          :tag="
            $t(
              'agent_builder.instructions.instructions_added.default_instruction',
            )
          "
        />
        <Instruction
          v-for="instruction in instructionsStore.instructions.data"
          :key="instruction.id"
          data-testid="instruction-added"
          :text="instruction.text"
        />
      </template>
    </section>
  </section>
</template>

<script setup>
import { computed } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

import Instruction from './Instruction.vue';
import i18n from '@/utils/plugins/i18n';

const instructionsStore = useInstructionsStore();
if (instructionsStore.instructions.status === null) {
  instructionsStore.loadInstructions();
}

const instructionsDefault = computed(() =>
  i18n.global.tm(
    'agent_builder.instructions.instructions_added.default_instructions',
  ),
);
</script>

<style lang="scss" scoped>
.instructions-added {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;

  &__title {
    margin: 0;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__instructions {
    display: flex;
    flex-direction: column;

    border-radius: $unnnic-border-radius-md;
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

    &--loading {
      gap: $unnnic-spacing-nano;

      border: none;
      border-radius: 0;
    }
  }
}
</style>
