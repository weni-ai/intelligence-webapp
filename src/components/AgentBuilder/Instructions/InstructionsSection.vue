<template>
  <section class="instructions-section">
    <h2
      class="instructions-section__title"
      data-testid="instructions-section-title"
    >
      {{ $t('agent_builder.instructions.instructions_list.title') }}
    </h2>

    <UnnnicTab
      :tabs="tabs"
      :activeTab="activeTab"
      data-testid="instructions-section-tab"
      @change="instructionsStore.activeInstructionsListTab = $event"
    >
      <template
        v-for="tab in tabs"
        :key="tab"
        #[`tab-head-${tab}`]
      >
        {{ $t(`agent_builder.instructions.instructions_list.tabs.${tab}`) }}
      </template>
      <template #[`tab-panel-${activeTab}`]>
        <p class="instructions-section__tab-description">
          {{
            $t(
              `agent_builder.instructions.instructions_list.tabs.${activeTab}_description`,
            )
          }}
        </p>
      </template>
    </UnnnicTab>
    <InstructionsList
      v-if="activeTab === 'custom'"
      data-testid="instructions-custom"
      :instructions="instructionsStore.instructions.data"
      :isLoading="instructionsStore.instructions.status === 'loading'"
      showActions
    />
    <InstructionsList
      v-if="activeTab === 'default'"
      data-testid="instructions-default"
      :instructions="instructionsDefault"
      :isLoading="false"
      :showActions="false"
    />
    <InstructionsList
      v-if="activeTab === 'safety_topics'"
      data-testid="instructions-safety-topics"
      :instructions="instructionsSafetyTopics"
      :isLoading="false"
      :showActions="false"
    />
  </section>
</template>

<script setup>
import { computed, ref } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

import InstructionsList from './InstructionsList.vue';
import i18n from '@/utils/plugins/i18n';

const instructionsStore = useInstructionsStore();

if (instructionsStore.instructions.status === null) {
  instructionsStore.loadInstructions();
}

const tabs = ref(['custom', 'default', 'safety_topics']);
const activeTab = computed(() => instructionsStore.activeInstructionsListTab);

function generateMockedInstructions(type) {
  return i18n.global
    .tm(`agent_builder.instructions.instructions_list.${type}`)
    .map((instruction, index) => ({
      id: `${type}-${index}`,
      text: instruction,
    }));
}

const instructionsDefault = computed(() =>
  generateMockedInstructions('default_instructions'),
);

const instructionsSafetyTopics = computed(() =>
  generateMockedInstructions('safety_topics'),
);
</script>

<style lang="scss" scoped>
.instructions-section {
  display: flex;
  flex-direction: column;

  &__title {
    margin: 0 0 $unnnic-spacing-sm;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__tab-description {
    margin-bottom: $unnnic-spacing-xs;

    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
