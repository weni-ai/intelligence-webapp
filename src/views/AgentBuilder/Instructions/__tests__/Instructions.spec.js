import { shallowMount } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';

import Instructions from '../index.vue';

describe('Instructions.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Instructions);
  });

  const SELECTORS = {
    instructionsHeader: '[data-testid="instructions-header"]',
    newInstruction: '[data-testid="new-instruction"]',
    instructionsSection: '[data-testid="instructions-section"]',
  };

  const findComponent = (component) =>
    wrapper.findComponent(SELECTORS[component]);

  describe('Component rendering', () => {
    it('should match snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders components correctly', () => {
      expect(findComponent('instructionsHeader').exists()).toBe(true);
      expect(findComponent('newInstruction').exists()).toBe(true);
      expect(findComponent('instructionsSection').exists()).toBe(true);
    });
  });
});
