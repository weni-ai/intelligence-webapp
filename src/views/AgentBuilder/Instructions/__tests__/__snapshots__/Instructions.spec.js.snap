// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Instructions.vue > Component rendering > should match snapshot 1`] = `
"<section data-v-d5e26ea0="" class="instructions">
  <agent-builder-header-stub data-v-d5e26ea0="" class="" withdivider="false" actionssize="none" data-testid="instructions-header"></agent-builder-header-stub>
  <new-instruction-stub data-v-d5e26ea0="" data-testid="new-instruction"></new-instruction-stub>
  <instructions-section-stub data-v-d5e26ea0="" data-testid="instructions-section"></instructions-section-stub>
</section>"
`;
